psi4/src/psi4/libscf_solver/sad.cc

```diff
@@ -358,61 +358,70 @@ void SADGuess::get_uhf_atomic_density(std::shared_ptr<BasisSet> bas, std::shared
     SharedMatrix Fa(mat.create_matrix("Fa"));
     SharedMatrix Fb(mat.create_matrix("Fb"));
 
-    // Factional occupation
+    // Fractional occupation
     SharedVector occ_a, occ_b;
-    if (options_.get_bool("SAD_FRAC_OCC")) {
-        int nfzc = 0, nact = 0;
-        if (Z <= 2) {
-            nfzc = 0;
-            nact = 1;
-        } else if (Z <= 4) {
-            nfzc = 1;
-            nact = 1;
-        } else if (Z <= 10) {
-            nfzc = 2;
-            nact = 3;
-        } else if (Z <= 18) {
-            nfzc = 5;
-            nact = 4;
-        } else if (Z <= 36) {
-            nfzc = 9;
-            nact = 9;
-        } else if (Z <= 54) {
-            nfzc = 18;
-            nact = 9;
-        } else if (Z <= 54) {
-            nfzc = 18;
-            nact = 9;
-        } else if (Z <= 86) {
-            nfzc = 27;
-            nact = 16;
+    // Number of partially or fully occupied orbitals
+    int nocc_a, nocc_b;
+    if (options_.get_bool("SAD_FRAC_OCC") || options_.get_bool("SAD_FRAC_SR_OCC")) {
+        // The density is spread over the whole of the possible valence shell.
+        // Only the noble gas core is doubly occupied
+        static const std::vector<int> magic_values = {0, 2, 10, 18, 36, 54, 86, 118};
+        // Shell index
+        size_t imagic = 0;
+        while(Z>magic_values[imagic+1]) {
+          imagic++;
+          if(imagic+1>=magic_values.size())
+            throw PSIEXCEPTION("SAD: Fractional occupations are not supported beyond Oganesson");
+        }
+
+        // Number of frozen orbitals
+        int nfzc = magic_values[imagic]/2;
+        // Number of active orbitals
+        int nact = magic_values[imagic+1]/2 - nfzc;
+
+        // Number of occupied orbitals is
+        nocc_a = nocc_b = nfzc + nact;
+        // Fractional alpha and beta occupation.
+        double frac_a, frac_b;
+
+        if (options_.get_bool("SAD_FRAC_SR_OCC")) {
+          // Spin-restricted occupations
+          frac_a = frac_b = (Z - 2.0*nfzc) / (2.0 * nact);
```

Comment:
Would you mind running `clang-format -i sad.cc` on this file, please?  We use four space indentation normally.

Reply:
Done.